crates/ide/src/completion/complete_postfix/format_like.rs

```diff
@@ -0,0 +1,325 @@
+                        self.output.push(chr);
+                        self.extracted_expressions.push(current_expr.trim().into());
+                        current_expr = String::new();
+                        self.state = State::FormatOpts;
+                    } else {
+                        // We're inside of braced expression, assume that it's a struct field name/value delimeter.
+                        current_expr.push(chr);
+                    }
+                }
+                (State::Expr, '{') => {
+                    current_expr.push(chr);
+                    inexpr_open_count += 1;
+                }
+                (State::Expr, _) => {
+                    current_expr.push(chr);
+                }
+                (State::FormatOpts, '}') => {
+                    self.output.push(chr);
+                    self.state = State::NotExpr;
+                }
+                (State::FormatOpts, _) => {
+                    self.output.push(chr);
+                }
+            }
+        }
+
+        if self.state != State::NotExpr {
+            return Err(());
+        }
+
+        self.parsed = true;
+        Ok(())
+    }
+
+    pub fn into_suggestion(&self, kind: PostfixKind) -> String {
+        assert!(self.parsed, "Attempt to get a suggestion from not parsed expression");
+
+        let mut output = format!(r#"{}("{}""#, kind.into_macro_name(), self.output);
+        for expr in &self.extracted_expressions {
+            output += ", ";
+            output += expr;
+        }
+        output.push(')');
+
+        output
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn format_str_parser() {
+        let test_vector = &[
+            ("no expressions", Some(("no expressions", vec![]))),
+            ("{expr} is {2 + 2}", Some(("{} is {}", vec!["expr", "2 + 2"]))),
+            ("{expr:?}", Some(("{:?}", vec!["expr"]))),
+            ("{malformed", None),
+            ("malformed}", None),
+            ("{{correct", Some(("{{correct", vec![]))),
+            ("correct}}", Some(("correct}}", vec![]))),
+            ("{correct}}}", Some(("{}}}", vec!["correct"]))),
+            ("{correct}}}}}", Some(("{}}}}}", vec!["correct"]))),
+            ("{incorrect}}", None),
+            ("placeholders {} {}", Some(("placeholders {} {}", vec!["$1", "$2"]))),
+            ("mixed {} {2 + 2} {}", Some(("mixed {} {} {}", vec!["$1", "2 + 2", "$2"]))),
+            (
+                "{SomeStruct { val_a: 0, val_b: 1 }}",
+                Some(("{}", vec!["SomeStruct { val_a: 0, val_b: 1 }"])),
+            ),
+            ("{expr:?} is {2.32f64:.5}", Some(("{:?} is {:.5}", vec!["expr", "2.32f64"]))),
+            (
+                "{SomeStruct { val_a: 0, val_b: 1 }:?}",
+                Some(("{:?}", vec!["SomeStruct { val_a: 0, val_b: 1 }"])),
+            ),
+            ("{     2 + 2        }", Some(("{}", vec!["2 + 2"]))),
+        ];
```

Comment:
To make this test even more data-driven, I suggest using an expect test:

```rust
let expected = expect![[r#"
no expressions
{} is {}; expr, 2 + 2
{:?}; expr
-
-
"#]];
```

It should be more resilient to changes in the `FormatStrParser` signature.